File: src/ext/circuit_generator.rs
//! Circuit generators
//!
//! Generate set of circuits based on some rules
//! Can define some exceptions?
//! Could be:
//! * Structured mirror circuits
//! * Random mirror circuits
//! * Cycle benchmarking
//! * Etc

use async_trait::async_trait;
use enum_dispatch::enum_dispatch;

use crate::circuit_generators::{
    BaseCircuitGenerator, BasicCircuitGenerator, FsCircuitGenerator, PythonCircuitGenerator,
    RandCircuitGenerator, StructCircuitGenerator, VolumeCircuitGenerator,
};
use crate::ext::LangSchemaDyn;
use crate::lang_schemas::LangCircuit;
use crate::{CliArgs, Error};

#[enum_dispatch]
pub enum CircuitGeneratorDyn {
    BasicCircuitGenerator,
    FsCircuitGenerator,
    VolumeCircuitGenerator,
    StructCircuitGenerator,
    RandCircuitGenerator,
    BaseCircuitGenerator,
    PythonCircuitGenerator,
}

#[async_trait]
#[enum_dispatch(CircuitGeneratorDyn)]
pub trait CircuitGenerator {
    /// Check whether arguments are correct
    fn check_constraints(&self, _args: &CliArgs) -> Result<(), Error> {
        Ok(())
    }

    async fn generate(
        &mut self,
        lang_schema: &LangSchemaDyn,
        i: i32,
        j: i32,
        rand: i32,
    ) -> Result<Option<LangCircuit>, Error>;
}

File: python/ext/circuit_generator.py
from enum import Enum
from typing import TypedDict, Optional


class LangGateType(str, Enum):
    X = 'X'


class LangGate(TypedDict):
    t: LangGateType
    i: int
    other: Optional[int]


class LangCircuit(TypedDict):
    width: int
    gates: list[LangGate]


class CircuitGenerator:
    def generate(self, i: int, j: int, it: int) -> LangCircuit:
        return {
            'width': 4,
            'gates': [{'t': LangGateType.X, 'i': 0, 'other': 42}]
        }
